### PR TITLE
*: fix server panic on invalid Election Proclaim/Resign HTTP requests

### PR DESCRIPTION
**Pull request by gyuho**
Server should never panic from wrong-formatted client requests. Without this patch, `/v3/election/proclaim` request with `leader` field missing can panic etcd servers.

@jpbetz This needs 3.2 backport as well.

### CODE REVIEW

**Comment by gyuho** [comment]
Fix https://github.com/coreos/etcd/issues/9375.
/cc @therve

**Comment by xiang90** [comment]
lgtm in general

File: etcdserver/api/v3election/election.go [hunk 1]
```diff
@@ -16,12 +16,17 @@ package v3election
...
+	"errors"
 
 	"github.com/coreos/etcd/clientv3"
 	"github.com/coreos/etcd/clientv3/concurrency"
 	epb "github.com/coreos/etcd/etcdserver/api/v3election/v3electionpb"
 )
 
+// ErrMissingLeaderField is returned when election API is
+// missing "leader" field.
+var ErrMissingLeaderField = errors.New(`missing "leader" field`)
```
[the remark this answers is not in the document]
Comment (gyuho) [reply etcdserver/api/v3election/election.go]: @xiang90 Right, we define as `LeaderKey` in proto. Just fixed.
Thanks!

File: etcdserver/api/v3election/election.go [hunk 2]
```diff
@@ -16,12 +16,17 @@ package v3election
...
+	"errors"
 
 	"github.com/coreos/etcd/clientv3"
 	"github.com/coreos/etcd/clientv3/concurrency"
 	epb "github.com/coreos/etcd/etcdserver/api/v3election/v3electionpb"
 )
 
+// ErrMissingLeaderKey is returned when election API is
+// missing "leader" field.
+var ErrMissingLeaderKey = errors.New(`missing "leader" field`)
```
Comment (xiang90) [inline etcdserver/api/v3election/election.go]: the error message also needs to be fixed.

Comment (gyuho) [reply etcdserver/api/v3election/election.go]: Oh I meant,

```go
type ProclaimRequest struct {
	// leader is the leadership hold on the election.
	Leader *LeaderKey `protobuf:"bytes,1,opt,name=leader" json:"leader,omitempty"`
	// value is an update meant to overwrite the leader's current value.
	Value []byte `protobuf:"bytes,2,opt,name=value,proto3" json:"value,omitempty"`
}
```

So, the JSON field is still `"leader"`

Comment (xiang90) [reply etcdserver/api/v3election/election.go]: oh. ok.

Comment (xiang90) [reply etcdserver/api/v3election/election.go]: the error message can be more descriptive like `leader field must be provided`

**Comment by xiang90** [comment]
lgtm

File: etcdserver/api/v3election/election.go [hunk 3]
```diff
@@ -16,12 +16,17 @@ package v3election
...
 	"context"
+	"errors"
 
 	"github.com/coreos/etcd/clientv3"
 	"github.com/coreos/etcd/clientv3/concurrency"
 	epb "github.com/coreos/etcd/etcdserver/api/v3election/v3electionpb"
 )
 
+// ErrMissingLeaderKey is returned when election API request
+// is missing "leader" field.
```
Comment (fanminshi) [inline etcdserver/api/v3election/election.go]: `missing "leader" field.` -> `missing the "leader" field.`?

**Comment by fanminshi** [comment]
lgtm

**Pull request merged**